File: darts-suite/c/darts-omp.c
/* Compute pi using OpenMP */
#include "lcgenerator.h"
#include <omp.h>
#include <stdio.h>

int main() {
    long num_trials = 1000000;
    long Ncirc = 0;
    double pi;
    double r = 1.0;
    double r2 = r*r;
    long base_seed = 12345;

#pragma omp parallel reduction(+:Ncirc)
    {
        long i;
        double x, y;
        int tid = omp_get_thread_num();
        int total_threads = omp_get_num_threads();
        long my_trials = num_trials / total_threads;

        for (i = 0; i < my_trials; i++) {
            long global_i = tid + (i * total_threads);
            long thread_seed = base_seed + (global_i * 2L);

            x = lcgrandom_r(&thread_seed);
            y = lcgrandom_r(&thread_seed);

            if ((x*x + y*y) <= r2) {
                Ncirc++;
            }
        }
    }

    pi = 4.0 * ((double)Ncirc) / ((double)num_trials);
    printf("\n \t Computing pi using OpenMP: \n");
    printf("\t For %ld trials, pi = %f\n", num_trials, pi);
    printf("\n");
    return 0;
}

File: darts-suite/fortran/darts-omp.f90
! Compute pi using OpenMP 
program darts
    use lcgenerator
    use omp_lib
    implicit none
    
    integer(kind=8) :: num_trials = 1000000_8
    integer(kind=8) :: ncirc = 0_8
    integer(kind=8) :: i, thread_seed, base_seed = 12345_8
    real(kind=8)    :: pi, x, y
    real(kind=8)    :: r = 1.0_8
    real(kind=8)    :: r2

    r2 = r * r

    !$omp parallel do private(thread_seed, x, y) reduction(+:ncirc)
    do i = 0, num_trials - 1
        ! Each iteration calculates its unique seed based on the loop index
        thread_seed = base_seed + (i * 2_8)
        
        call lcgrandom_r(thread_seed, x)
        call lcgrandom_r(thread_seed, y)
        
        if ((x*x + y*y) <= r2) then
            ncirc = ncirc + 1_8
        end if
    end do
    !$omp end parallel do

    pi = 4.0_8 * real(ncirc, 8) / real(num_trials, 8)
    print*, '     '
    print*, '     Computing pi using OpenMP:'
    print '(/,A,I0,A,F10.6,/)', '      For ', num_trials, ' trials, pi = ', pi
    print*, '     '     
end program
